File: modules/battery-level/ios/BatteryGaugeView.swift
import ExpoModulesCore
import UIKit

/**
 A presentational battery pill drawn with Core Graphics. It holds no battery
 state of its own — the level is passed in as a prop, so the module stays the
 single source of truth and the view stays trivially testable from JS.
 */
class BatteryGaugeView: ExpoView {
  /// Matched by name to `Events("onPress")` in the view definition.
  let onPress = EventDispatcher()

  var level: Float = -1
  var isCharging = false
  var fillColor: UIColor = .systemGreen
  var trackColor: UIColor = .systemGray3
  var boltColor: UIColor = .white

  required init(appContext: AppContext? = nil) {
    super.init(appContext: appContext)
    backgroundColor = .clear
    isOpaque = false
    addGestureRecognizer(UITapGestureRecognizer(target: self, action: #selector(handleTap)))
  }

  /// Called once from `OnViewDidUpdateProps`, so a batch of prop changes
  /// results in a single redraw rather than one per prop.
  func redraw() {
    setNeedsDisplay()
  }

  @objc private func handleTap() {
    onPress(["level": level])
  }

  override func draw(_ rect: CGRect) {
    let lineWidth: CGFloat = 2
    let capWidth = max(rect.width * 0.05, 2)
    let shell = CGRect(
      x: rect.minX + lineWidth / 2,
      y: rect.minY + lineWidth / 2,
      width: rect.width - capWidth - 2 - lineWidth,
      height: rect.height - lineWidth
    )
    let radius = shell.height * 0.32

    // Outline
    let outline = UIBezierPath(roundedRect: shell, cornerRadius: radius)
    outline.lineWidth = lineWidth
    trackColor.setStroke()
    outline.stroke()

    // Terminal nub on the right
    let capHeight = shell.height * 0.4
    let cap = UIBezierPath(
      roundedRect: CGRect(
        x: shell.maxX + 2,
        y: shell.midY - capHeight / 2,
        width: capWidth,
        height: capHeight
      ),
      cornerRadius: capWidth / 2
    )
    trackColor.setFill()
    cap.fill()

    // An unknown level (-1) draws the empty shell only.
    guard level >= 0 else { return }

    let track = shell.insetBy(dx: lineWidth + 1, dy: lineWidth + 1)
    let clamped = CGFloat(min(max(level, 0), 1))
    let fillWidth = track.width * clamped

    if fillWidth > 0 {
      let fill = UIBezierPath(
        roundedRect: CGRect(
          x: track.minX,
          y: track.minY,
          // Keep a sliver visible at very low levels.
          width: max(fillWidth, track.height * 0.4),
          height: track.height
        ),
        cornerRadius: radius * 0.6
      )
      fillColor.setFill()
      fill.fill()
    }

    if isCharging {
      drawBolt(in: track)
    }
  }

  private func drawBolt(in rect: CGRect) {
    let height = rect.height * 0.7
    let width = height * 0.5
    let x = rect.midX - width / 2
    let y = rect.midY - height / 2

    let bolt = UIBezierPath()
    bolt.move(to: CGPoint(x: x + width * 0.55, y: y))
    bolt.addLine(to: CGPoint(x: x, y: y + height * 0.58))
    bolt.addLine(to: CGPoint(x: x + width * 0.42, y: y + height * 0.58))
    bolt.addLine(to: CGPoint(x: x + width * 0.30, y: y + height))
    bolt.addLine(to: CGPoint(x: x + width, y: y + height * 0.40))
    bolt.addLine(to: CGPoint(x: x + width * 0.55, y: y + height * 0.40))
    bolt.close()

    boltColor.setFill()
    bolt.fill()
  }
}

File: modules/battery-level/android/src/main/java/expo/modules/batterylevel/BatteryGaugeView.kt
package expo.modules.batterylevel

import android.content.Context
import android.graphics.Canvas
import android.graphics.Color
import android.graphics.Paint
import android.graphics.Path
import android.graphics.RectF
import expo.modules.kotlin.AppContext
import expo.modules.kotlin.viewevent.EventDispatcher
import expo.modules.kotlin.views.ExpoView

/**
 * Canvas counterpart of the iOS BatteryGaugeView. It holds no battery state of
 * its own — the level arrives as a prop, so the module stays the single source
 * of truth.
 */
class BatteryGaugeView(context: Context, appContext: AppContext) : ExpoView(context, appContext) {
  /** Matched by name to `Events("onPress")` in the view definition. */
  private val onPress by EventDispatcher()

  var level: Float = -1f
  var isCharging: Boolean = false
  var fillColor: Int = Color.GREEN
  var trackColor: Int = Color.GRAY
  var boltColor: Int = Color.WHITE

  private val strokePaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.STROKE
  }
  private val fillPaint = Paint(Paint.ANTI_ALIAS_FLAG).apply {
    style = Paint.Style.FILL
  }
  private val boltPath = Path()

  init {
    // ExpoView is a ViewGroup, which skips onDraw unless this is cleared.
    setWillNotDraw(false)
    isClickable = true
    setOnClickListener {
      onPress(mapOf("level" to level))
    }
  }

  /**
   * Called once from `OnViewDidUpdateProps`, so a batch of prop changes results
   * in a single redraw rather than one per prop.
   */
  fun redraw() {
    invalidate()
  }

  override fun onDraw(canvas: Canvas) {
    super.onDraw(canvas)

    val strokeWidth = 2f * resources.displayMetrics.density
    val capWidth = (width * 0.05f).coerceAtLeast(strokeWidth)

    val shell = RectF(
      strokeWidth / 2f,
      strokeWidth / 2f,
      width - capWidth - strokeWidth * 1.5f,
      height - strokeWidth / 2f
    )
    val radius = shell.height() * 0.32f

    strokePaint.color = trackColor
    strokePaint.strokeWidth = strokeWidth
    canvas.drawRoundRect(shell, radius, radius, strokePaint)

    // Terminal nub on the right
    val capHeight = shell.height() * 0.4f
    val cap = RectF(
      shell.right + strokeWidth,
      shell.centerY() - capHeight / 2f,
      shell.right + strokeWidth + capWidth,
      shell.centerY() + capHeight / 2f
    )
    fillPaint.color = trackColor
    canvas.drawRoundRect(cap, capWidth / 2f, capWidth / 2f, fillPaint)

    // An unknown level (-1) draws the empty shell only.
    if (level < 0f) {
      return
    }

    val inset = strokeWidth + 1f
    val track = RectF(
      shell.left + inset,
      shell.top + inset,
      shell.right - inset,
      shell.bottom - inset
    )
    val clamped = level.coerceIn(0f, 1f)
    val fillWidth = track.width() * clamped

    if (fillWidth > 0f) {
      val fill = RectF(
        track.left,
        track.top,
        // Keep a sliver visible at very low levels.
        track.left + fillWidth.coerceAtLeast(track.height() * 0.4f),
        track.bottom
      )
      fillPaint.color = fillColor
      canvas.drawRoundRect(fill, radius * 0.6f, radius * 0.6f, fillPaint)
    }

    if (isCharging) {
      drawBolt(canvas, track)
    }
  }

  private fun drawBolt(canvas: Canvas, rect: RectF) {
    val h = rect.height() * 0.7f
    val w = h * 0.5f
    val x = rect.centerX() - w / 2f
    val y = rect.centerY() - h / 2f

    boltPath.reset()
    boltPath.moveTo(x + w * 0.55f, y)
    boltPath.lineTo(x, y + h * 0.58f)
    boltPath.lineTo(x + w * 0.42f, y + h * 0.58f)
    boltPath.lineTo(x + w * 0.30f, y + h)
    boltPath.lineTo(x + w, y + h * 0.40f)
    boltPath.lineTo(x + w * 0.55f, y + h * 0.40f)
    boltPath.close()

    fillPaint.color = boltColor
    canvas.drawPath(boltPath, fillPaint)
  }
}
